core/trino-main/src/main/java/io/trino/operator/window/matcher/IntListMap.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.window.matcher;
+
+import io.airlift.slice.SizeOf;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+class IntListMap
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(IntListMap.class).instanceSize();
+
+    private IntList[] values;
+    private final int listCapacity;
+
+    public IntListMap(int capacity, int listCapacity)
+    {
+        this.values = new IntList[capacity];
+        this.listCapacity = listCapacity;
+    }
+
+    public void add(int key, int value)
+    {
+        ensureCapacity(key);
+        if (values[key] == null) {
+            values[key] = new IntList(listCapacity);
+        }
+        values[key].add(value);
+    }
+
+    private void put(int key, IntList valueList)
+    {
+        ensureCapacity(key);
+        values[key] = valueList;
+    }
+
+    public ArrayView getArrayView(int key)
+    {
+        if (values[key] == null) {
+            return new ArrayView(new int[] {}, 0);
+        }
+        return values[key].toArrayView();
+    }
+
+    public void release(int key)
+    {
+        put(key, null);
```

Comment:
ensureCapacity shouldn't be needed in this case. I would specialize to `values[key] = null`.